File: backend/src/main/java/com/yibang/erp/domain/entity/Product.java
package com.yibang.erp.domain.entity;

import com.baomidou.mybatisplus.annotation.*;
import com.yibang.erp.domain.enums.ProductApprovalStatus;
import com.yibang.erp.domain.enums.ProductStatus;
import lombok.Data;
import lombok.EqualsAndHashCode;

import java.math.BigDecimal;
import java.time.LocalDateTime;

/**
 * 商品实体类
 *
 * @author yibang-erp
 * @since 2024-01-14
 */
@Data
@EqualsAndHashCode(callSuper = false)
@TableName("products")
public class Product {

    /**
     * 商品ID
     */
    @TableId(value = "id", type = IdType.AUTO)
    private Long id;

    /**
     * 商品编码
     */
    @TableField("sku")
    private String sku;

    /**
     * 商品名称
     */
    @TableField("name")
    private String name;

    /**
     * 分类ID
     */
    @TableField("category_id")
    private Long categoryId;

    /**
     * 品牌ID
     */
    @TableField("brand_id")
    private Long brandId;

    /**
     * 商品描述
     */
    @TableField("description")
    private String description;

    /**
     * 商品简介
     */
    @TableField("short_description")
    private String shortDescription;

    /**
     * 商品图片JSON数组
     */
    @TableField("images")
    private String images;

    /**
     * 商品规格JSON
     */
    @TableField("specifications")
    private String specifications;

    /**
     * 单位
     */
    @TableField("unit")
    private String unit;

    /**
     * 成本价
     */
    @TableField("cost_price")
    private BigDecimal costPrice;

    /**
     * 销售价
     */
    @TableField("selling_price")
    private BigDecimal sellingPrice;

    /**
     * 市场价
     */
    @TableField("market_price")
    private BigDecimal marketPrice;

    /**
     * 重量(kg)
     */
    @TableField("weight")
    private BigDecimal weight;

    /**
     * 尺寸信息JSON
     */
    @TableField("dimensions")
    private String dimensions;

    /**
     * 条形码
     */
    @TableField("barcode")
    private String barcode;

    /**
     * 海关编码
     */
    @TableField("hs_code")
    private String hsCode;

    /**
     * 原产国
     */
    @TableField("origin_country")
    private String originCountry;

    /**
     * 材质
     */
    @TableField("material")
    private String material;

    /**
     * 颜色
     */
    @TableField("color")
    private String color;

    /**
     * 尺寸
     */
    @TableField("size")
    private String size;

    /**
     * 商品标签JSON数组
     */
    @TableField("tags")
    private String tags;

    /**
     * 商品状态
     */
    @TableField("status")
    private ProductStatus status;

    /**
     * 审核状态
     */
    @TableField("approval_status")
    private ProductApprovalStatus approvalStatus;

    /**
     * 审核意见
     */
    @TableField("approval_comment")
    private String approvalComment;

    /**
     * 审核时间
     */
    @TableField("approval_at")
    private LocalDateTime approvalAt;

    /**
     * 审核人ID
     */
    @TableField("approval_by")
    private Long approvalBy;

    /**
     * 所属公司ID
     */
    @TableField("company_id")
    private Long companyId;

    /**
     * 是否推荐
     */
    @TableField("is_featured")
    private Boolean isFeatured;

    /**
     * 是否热销
     */
    @TableField("is_hot")
    private Boolean isHot;

    /**
     * 是否新品
     */
    @TableField("is_new")
    private Boolean isNew;

    /**
     * 浏览次数
     */
    @TableField("view_count")
    private Long viewCount;

    /**
     * 创建时间
     */
    @TableField(value = "created_at", fill = FieldFill.INSERT)
    private LocalDateTime createdAt;

    /**
     * 更新时间
     */
    @TableField(value = "updated_at", fill = FieldFill.INSERT_UPDATE)
    private LocalDateTime updatedAt;

    /**
     * 创建人ID
     */
    @TableField("created_by")
    private Long createdBy;

    /**
     * 更新人ID
     */
    @TableField("updated_by")
    private Long updatedBy;

    /**
     * 逻辑删除标记
     */
    @TableLogic
    @TableField("deleted")
    private Boolean deleted;

    /**
     * 检查商品是否可以编辑
     */
    public boolean isEditable() {
        return status != null && status.isEditable();
    }

    /**
     * 检查商品是否可以销售
     */
    public boolean isSellable() {
        return status != null && status.isSellable();
    }

    /**
     * 检查商品是否已审核
     */
    public boolean isApproved() {
        return approvalStatus != null && approvalStatus.isApproved();
    }

    /**
     * 检查商品是否被拒绝
     */
    public boolean isRejected() {
        return approvalStatus != null && approvalStatus.isRejected();
    }

    /**
     * 检查商品是否需要审核
     */
    public boolean needsApproval() {
        return approvalStatus == null || approvalStatus == ProductApprovalStatus.PENDING;
    }
}

File: frontend/src/types/product.ts
/**
 * 商品相关类型定义
 */

// 商品状态枚举
export enum ProductStatus {
  DRAFT = 'DRAFT',
  PENDING = 'PENDING',
  ACTIVE = 'ACTIVE',
  INACTIVE = 'INACTIVE',
  DISCONTINUED = 'DISCONTINUED'
}

// 商品审核状态枚举
export enum ProductApprovalStatus {
  PENDING = 'PENDING',
  APPROVED = 'APPROVED',
  REJECTED = 'REJECTED'
}

// 商品实体类型
export interface Product {
  id: number
  sku: string
  name: string
  categoryId: number
  brandId?: number
  description?: string
  shortDescription?: string
  images?: string
  specifications?: string
  unit: string
  costPrice: number
  sellingPrice: number
  marketPrice?: number
  retailLimitPrice?: number
  weight?: number
  dimensions?: string
  barcode?: string
  hsCode?: string
  originCountry?: string
  material?: string
  color?: string
  size?: string
  tags?: string
  status: ProductStatus
  approvalStatus: ProductApprovalStatus
  approvalComment?: string
  approvalAt?: string
  approvalBy?: number
  companyId: number
  isFeatured: boolean
  isHot: boolean
  isNew: boolean
  viewCount: number
  createdAt: string
  updatedAt: string
  createdBy?: number
  updatedBy?: number
  deleted: boolean
  priceTierConfigs?: ProductPriceTierConfig[] // 添加价格分层配置
}

// 商品创建请求类型
export interface ProductCreateRequest {
  sku: string
  name: string
  categoryId: number
  brandId?: number
  description?: string
  shortDescription?: string
  unit: string
  costPrice: number
  sellingPrice: number
  marketPrice?: number
  retailLimitPrice?: number
  weight?: number
  material?: string
  color?: string
  size?: string
  originCountry?: string
  hsCode?: string
  tags?: string
  isFeatured?: boolean
  isHot?: boolean
  isNew?: boolean
  companyId: number        // 添加公司ID（必填）
  approvalStatus?: string  // 添加审核状态
}

// 商品更新请求类型
export interface ProductUpdateRequest extends Partial<ProductCreateRequest> {
  id: number
}

// 商品查询请求类型
export interface ProductQueryRequest {
  page: number
  size: number
  name?: string
  status?: string
  approvalStatus?: string
  categoryId?: number
  brandId?: number
  companyId?: number
}

// 商品审核请求类型
export interface ProductAuditRequest {
  productId: number
  approvalStatus: ProductApprovalStatus
  approvalComment?: string
  approverId: number
}

// 商品审核结果类型
export interface ProductAuditResult {
  productId: number
  productName: string
  sku: string
  approvalStatus: ProductApprovalStatus
  productStatus: ProductStatus
  approvalComment?: string
  approverId: number
  approverName?: string
  approvalAt: string
  message: string
  success: boolean
}

// 商品分页结果类型
export interface ProductPageResult {
  records: Product[]
  total: number
  size: number
  current: number
  pages: number
}

// 商品分类类型
export interface ProductCategory {
  id: number
  name: string
  parentId?: number
  level: number
  sortOrder: number
  description?: string
  icon?: string
  status: string
  createdAt: string
  updatedAt: string
}

// 商品品牌类型
export interface ProductBrand {
  id: number
  name: string
  logo?: string
  description?: string
  website?: string
  status: string
  createdAt: string
  updatedAt: string
}

// 商品库存类型
export interface ProductInventory {
  id: number
  productId: number
  warehouseId: number
  availableQuantity: number
  reservedQuantity: number
  damagedQuantity: number
  minStockLevel: number
  maxStockLevel?: number
  reorderPoint?: number
  lastStockIn?: string
  lastStockOut?: string
  createdAt: string
  updatedAt: string
}

// 商品价格类型
export interface ProductPrice {
  id: number
  productId: number
  priceType: string
  price: number
  currency: string
  effectiveFrom: string
  effectiveTo?: string
  minQuantity?: number
  maxQuantity?: number
  customerGroup?: string
  companyId?: number
  createdAt: string
  updatedAt: string
}

// 商品审核日志类型
export interface ProductApprovalLog {
  id: number
  productId: number
  action: string
  status: string
  comment?: string
  approverId?: number
  approvalAt: string
  createdAt: string
}

// 商品统计信息类型
export interface ProductStatistics {
  totalCount: number
  draftCount: number
  pendingCount: number
  activeCount: number
  inactiveCount: number
  discontinuedCount: number
  approvedCount: number
  rejectedCount: number
  featuredCount: number
  hotCount: number
  newCount: number
}

// 商品价格分层配置类型
export interface ProductPriceTierConfig {
  id: number
  productId: number
  priceTierId: number
  priceTierName: string
  priceTierType: string
  dropshippingPrice: number // 一件代发价格
  retailLimitPrice: number // 零售限价
  isActive: boolean
  createdAt: string
  updatedAt: string
}

// 商品价格分层配置请求类型
export interface ProductPriceTierConfigRequest {
  productId: number
  priceTierId: number
  dropshippingPrice: number
  retailLimitPrice: number
  isActive: boolean
  companyId?: number    // 添加公司ID
  createdBy?: number    // 添加创建人ID
}

// 商品价格分层配置响应类型
export interface ProductPriceTierConfigResponse {
  id: number
  productId: number
  priceTierId: number
  priceTierName: string
  priceTierType: string
  dropshippingPrice: number
  retailLimitPrice: number
  isActive: boolean
  createdAt: string
  updatedAt: string
}

// 商品价格配置查询请求类型
export interface ProductPriceConfigQueryRequest {
  productId: number
  companyId?: number
  priceTierId?: number
  isActive?: boolean
}
